libp2p/multistream.nim

```diff
@@ -242,8 +260,8 @@ proc addHandler*(m: MultistreamSelect,
                                protocol: protocol,
                                match: matcher))
 
-proc start*(m: MultistreamSelect) {.async.} =
-  await allFutures(m.handlers.mapIt(it.protocol.start()))
+proc start*(m: MultistreamSelect) {.async: (raises: []).} =
+  await noCancel allFutures(m.handlers.mapIt(it.protocol.start()))
```

Comment:
hm, this is somewhat unexpected, that protocol start functions shouldn't support cancellation - I'm thinking about the case where for example a protocol might need to perform a dns lookup or the like and we want to exit the process - timeouts in such protocols can be longer than the "cancellation tolerance".

Reply:
Once more it is simply reflecting current behaviour. But agree that having cancellation on `start` feels more intuitive. Have updated the logic to support that, and also to properly process cancellation to avoid inconsistent state.

Reply:
When deciding how to constrain raises, it's important to take into account an important aspect of exception-based API: when they raise `CatchableError` it means the calling code must be prepared to handle any error _even if_ de facto there's no raise today - this is an important "feature" of designing exception-based api: you can "future-proof" them by signalling in the signature that you're raising more than you actually are - this is why the strategy of simply "describing current behavior" is .. incomplete: it doesn't take into account the non-code intent that went into the design.

Of course, in this particular case, there was probably no intent or thinking about cancellation at all since the code didn't actually do cancellation correctly until your fix, but when "narrowing" an API to raise less, it's a good question to ask.

Reply:
Yes, widening it to match the intended design space makes sense.